strawberryfields/apps/sample.py

```diff
@@ -361,3 +377,66 @@ def vibronic(
     if np.any(t == 0):
         s = np.pad(s, ((0, 0), (0, n_modes))).tolist()
     return s
+
+
+def waw_matrix(A: np.ndarray, w: Union[np.ndarray, list]) -> np.ndarray:
+    r"""Rescale adjacency matrix to account for node weights.
+
+    Given a graph with adjacency matrix :math:`A` and a vector :math:`w` of weighted nodes,
+    this function rescales the adjacency matrix according to:
+
+    .. math::
+
+        A \rightarrow WAW,
+
+    with :math:`W` the diagonal matrix formed by the weighted nodes. This encoding can be used to
+    add a bias in GBS toward selecting subgraphs with a large product of node weights
+    :cite:`banchi2019molecular`. The resulting matrix can be passed to :func:`sample`.
+
+    **Example usage:**
+
+    >>> g = nx.erdos_renyi_graph(5, 0.7)
+    >>> a = nx.to_numpy_array(g)
+    >>> w = [1, 1, 3, 1, 0.5]
+    >>> a = waw_matrix(a, w)
+    >>> sample(a, 3, 4)
+    [[0, 0, 0, 0, 0], [1, 0, 0, 0, 1], [0, 0, 0, 0, 0], [0, 1, 1, 1, 1]]
+
+    Args:
+        A (array): adjacency matrix to rescale
+        w (array or list): vector of real node weights
+
+    Returns:
+        array: matrix rescaled according to the WAW encoding
+    """
+    if not np.allclose(A, A.T):
+        raise ValueError("Input must be a NumPy array corresponding to a symmetric matrix")
+
+    w = _process_w(len(A), w)
+
+    return (w * A).T * w
+
+
+def _process_w(dim: int, w: Union[np.ndarray, list]) -> np.ndarray:
```

Comment:
I don't mind this function, but isn't it a bit of overkill?

Reply:
Is the idea that we are allowing the weight vector to have different forms? Isn't it better to have a fixed form?

Reply:
The user might want to input the weight vector as:
- a list
- a numpy array of shape (dim,) - standard
- a numpy array of shape (1, dim) - a 2d row vector
- a numpy array of shape (dim, 1) - a 2d column vector

Actually the above line 417 is fine with any of them... so maybe you're right and we can skip the checking. Although one thing is we definitely should let the user input as a list or numpy array.

Reply:
Ok, simplified